core/camel-support/src/main/java/org/apache/camel/support/AbstractExchange.java

```diff
@@ -886,4 +912,32 @@ public String toString() {
         }
     }
 
+    @Override
+    public void setCopySafeProperty(String key, CamelCopySafeProperty<?> value) {
+        if (value != null) {
+            // avoid the NullPointException
+            if (copySafeProperties == null) {
+                this.copySafeProperties = new ConcurrentHashMap<>(2);
+            }
+            copySafeProperties.put(key, value);
+        } else if (copySafeProperties != null) {
+            // if the value is null, we just remove the key from the map
+            copySafeProperties.remove(key);
+        }
+
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> T getCopySafeProperty(String key, Class<T> type) {
+
+        Object value = getCopySafeProperties().get(key);
```

Comment:
Optimize to check if there is any properties at all (ie its null) to then return null quickly

Reply:
Thanks @davsclaus . Updated the PR based on your feedback.